basic networking

tweak-wtf opened the pull request:
## Changelog Description
Adds basic support for loading representations into a network-managed OpenRV session from TrayLoader.
It uses OpenRVs networking feature to send custom events from any network source to a running instance of OpenRV.

- adds `Open in RV` Loader action
- implements `ayon_load_container` RV event
- adds networking and imageio server settings

## Additional Info
Your need to launch RV with `-network` in the host settings.
![image](https://github.com/ynput/ayon-openrv/assets/4457962/6eccc6e5-df76-4c0c-b218-5ad204115965)
`ayon+settings://applications/applications/openrv`

Unfortunately I know of no way to launch RV with a specific port. If you launch multiple instances of RV with `-network` they will each increment their socket port. Depending on the order of closing the instances RV might remember the "wrong" port resulting in further pings to fail. In that case either reset RV's networking to default or set the new port in server settings.
The server network settings default to RV's
When using `Open in RV` to launch with multiple representations it may happen that loading the media exceeds the ping timeout. If you already have an open session you should be fine.

## Testing Notes
1. open TrayLoader
2. select supported products
3. click Open In RV

BigRoy commented on server/settings/main.py:
```diff
@@ -1,14 +1,17 @@
-from ayon_server.settings import BaseSettingsModel
-
-DEFAULT_VALUES = {}
+from .imageio import ImageIOSettings, DEFAULT_IMAGEIO_SETTINGS
```
`DEFAULT_IMAGEIO_SETTINGS` seems unused. By accident?

tweak-wtf replied:
u are completely right!

tweak-wtf replied:
removed it

BigRoy commented on server/settings/imageio.py:
```diff
@@ -0,0 +1,43 @@
+""" Providing models and setting values for image IO in OpenRV.
+    Copied from maya's settings
+"""
```
```suggestion
"""Providing models and setting values for image IO in OpenRV."""
```

In the end we don't care what it was based on.

tweak-wtf replied:
that's done

BigRoy commented on client/ayon_openrv/startup/pkgs_source/ayon_menus/ayon_menus.py:
```diff
@@ -41,36 +47,36 @@ def install_openpype_to_host():
+            "",
+            None,
+            [
+                ("ayon_open_loader", self.load, "Opens AYON Loader."),
+                ("ayon_load_container", on_ayon_load_container, "Loads an AYON representation into the session.")
             ],
-            # initialization order
-            sortKey="source_setup",
-            ordering=15
+            self.build_menu(),
```
Can we keep the argument names in there? `name=`, `globalBindings=`, etc. makes it much easier to understand what's going on.

tweak-wtf replied:
sure thing. i'll check out the signature of that function of `rv.MinorMode`

BigRoy replied:
Even if positional arguments you should still be able to pass them by name, right?

[Should be fine with code like this](https://github.com/AcademySoftwareFoundation/OpenRV/blob/dce1b6246fdca4491ee66a0f09f438aafde5e0c2/src/lib/app/py_rvui/rv/rvtypes.py#L135-L137) unless they would've been using something like `*args` right?

```python
def tony(level, upgrade=False):
    if upgrade:
        level +=1
    print(f"Tony is at level {level}")
    if level > 9000:
        print("IT'S OVER 9000!!!!!")
    
tony(level=9000, upgrade=True)
# Tony is at level 9001
# IT'S OVER 9000!!!!!
```

tweak-wtf replied:
oh true, that's cool! o.O
didn't know that. pushed ✌️

BigRoy commented on client/ayon_openrv/plugins/load/load_frames.py:
```diff
@@ -55,8 +55,8 @@ def load(self, context, name=None, namespace=None, data=None):
 
     def update(self, container, representation):
```
This should be 
```suggestion
    def update(self, container, context):
```
In recent versions of ayon-core. Depends on which version you're still working with, likely the `openpype` base, but you're also refactoring to `ayon_core` so I'm not sure? :)

tweak-wtf replied:
and that's refactored too now

BigRoy commented on client/ayon_openrv/plugins/load/load_frames.py:
```diff
@@ -55,8 +55,8 @@ def load(self, context, name=None, namespace=None, data=None):
         node = container["node"]
-
-        context = get_representation_context(representation)
+        repr_id = representation["representation"]["_id"]
```
```suggestion
        repr_id = representation["representation"]["id"]
```

Ids in ayon api are not prefixed with `_` like they were in mongodb

tweak-wtf replied:
ok cool. i'll try it out.
was just printing out `representation` to see whats available to me and only found `_id` key in the dict.

BigRoy replied:
Interesting - then you might be on older ayon core still. If so, then ignore me. I believe we're about just under a week out of a potential ayon core 0.3 release which should make that the new public standard version. @iLLiCiTiT will hit me for giving such a release ETA.

tweak-wtf replied:
yeah i just hit the KeyError on `_id`
was using quite an outdated ayon_core. just merged everything in and currently fixing the integration to work with latest ayon_core

tweak-wtf replied:
it actually trashed quite a lot :D
i'll try to file a PR next week but till then i pushed to my fork.
it's here in case
https://github.com/ynput/ayon-core/compare/develop...tweak-wtf:ayon-core:bpc/openrv-networking

BigRoy commented on client/ayon_openrv/startup/pkgs_source/ayon_menus/ayon_menus.py:
```diff
@@ -101,12 +107,58 @@ def data_loader():
+
+    # separate generic nodes from ayon containers
+    generic_nodes = []
+    ayon_containers = {"FramesLoader": [], "MovLoader": []}
+    for node in event_contents:
+        if node.get("file"):
+            generic_nodes.append(node)
+            continue
+
+        if node.get("representation"):
+            for ext in IMAGE_EXTENSIONS:
+                ext = ext.lstrip(".")
+                if ext in node["objectName"].lower():
+                    ayon_containers["FramesLoader"].append(node)
+                    break
+            for ext in VIDEO_EXTENSIONS:
+                ext = ext.lstrip(".")
+                if ext in node["objectName"].lower():
+                    ayon_containers["MovLoader"].append(node)
+                    break
+
+    if ayon_containers["FramesLoader"]:
+        # load the container with appropriate loader plugin
+        # this enables us to use AYON manager for versioning
+        representation_ids = [i["representation"] for i in ayon_containers["FramesLoader"]]
+        log.debug(f"{representation_ids = }")
+        load_data(dataset=representation_ids, loader_type="FramesLoader")
+
+    if ayon_containers["MovLoader"]:
+        # load the container with appropriate loader plugin
+        # this enables us to use AYON manager for versioning
+        representation_ids = [i["representation"] for i in ayon_containers["MovLoader"]]
+        log.debug(f"{representation_ids = }")
+        load_data(dataset=representation_ids, loader_type="MovLoader")
+    
+    if generic_nodes:
+        # load them "normally" via networking or rv.commands
+        source_paths = list([n["file"] for n in generic_nodes])
+        log.debug(f"{source_paths = }")
+        rv.commands.addSourceVerbose(source_paths)
```
It might be nice to move this logic outside of the `pkgs_source` menus file and instead just call a function from within the `ayon_openrv/api` somewhere so that the logic itself that it triggers is defined there, keeping the `ayon_menus.py` lightweight.

BigRoy commented on client/ayon_openrv/startup/pkgs_source/ayon_scripteditor/ayon_scripteditor.py:
```diff
@@ -10,7 +10,7 @@
 try:
-    from openpype_modules.python_console_interpreter.window import PythonInterpreterWidget  # noqa
+    from ayon_core.modules.python_console_interpreter.window import PythonInterpreterWidget  # noqa
 except ModuleNotFoundError:
     from openpype.modules.python_console_interpreter.window import PythonInterpreterWidget  # noqa
```
Pretty sure this legacy fallback can be removed?

tweak-wtf replied:
think so too. just didn't want to remove openpype since i believe back then ynput was still maintaining OpenPype repo.
so is this safe to drop support for openpype then?

BigRoy replied:
I believe `ayon-openrv` is never intended to be compatible with OpenPype. As such, it can be removed. @iLLiCiTiT can you confirm?

jakubjezek001 replied:
yes please, all `openpype` imports can be replaced for `ayon-core`.

iLLiCiTiT replied:
I would say that should happen in different PR.

antirotor replied:
> I would say that should happen in different PR.

I would do it here if [this](https://github.com/ynput/ayon-openrv/pull/1/files#r1548542392) is also taken care of.

BigRoy commented on client/ayon_openrv/plugins/load/load_mov.py:
```diff
@@ -65,7 +65,7 @@ def update(self, container, representation):
                                       ["repname"], True)
         rv.commands.setStringProperty(node + ".openpype.representation",
-                                      [str(representation["_id"])], True)
+                                      [str(repr_id)], True)
```
The `str()` conversion should not be needed anymore. In AYON the representation id is always a string.
```suggestion
                                      [repr_id], True)
```

BigRoy commented on client/ayon_openrv/plugins/load/load_frames.py:
```diff
@@ -72,7 +72,7 @@ def update(self, container, representation):
                                       ["repname"], True)
         rv.commands.setStringProperty(node + ".openpype.representation",
-                                      [str(representation["_id"])], True)
+                                      [str(repr_id)], True)
```
The `str()` conversion should not be needed anymore. In AYON the representation id is always a string.
```suggestion
                                      [repr_id], True)
```

BigRoy commented on client/ayon_openrv/plugins/load/load_mov.py:
```diff
@@ -48,8 +48,8 @@ def load(self, context, name=None, namespace=None, data=None):
 
     def update(self, container, representation):
```
This `representation` argument should be `context`
```suggestion
    def update(self, container, context):
```

Which likely explains better for you also the different behavior below. ;)

BigRoy replied:
Same as what was mentioned [here](https://github.com/ynput/ayon-openrv/pull/1/files?diff=unified&w=0#r1542756446)

tweak-wtf replied:
that's refactored now

BigRoy commented on client/ayon_openrv/plugins/load/load_mov.py:
```diff
@@ -48,8 +48,8 @@ def load(self, context, name=None, namespace=None, data=None):
         node = container["node"]
-
-        context = get_representation_context(representation)
+        repr_id = representation["representation"]["_id"]
+        context = get_representation_context(representation["representation"])
         filepath = load.get_representation_path_from_context(context)
```
```suggestion
        repr_id = context["representation"]["_id"]
        filepath = self.filepath_from_context(context)
```

BigRoy commented on client/ayon_openrv/api/pipeline.py:
```diff
@@ -4,11 +4,15 @@
-import rv
 
-from openpype.host import HostBase, ILoadHost, IWorkfileHost, IPublishHost
+try:
+    import rv
+except ImportError:
+    pass
```
Why is this needed?

tweak-wtf replied:
oh yeah that's a flaw in the RvConnector import.
it tries to import rv when right-clicking something in TrayLoader. gotta fix this in `client\ayon_openrv\api\networking.py` so i can remove this try/catch  

OT: PR for ayon_core comes next week. there are more changes to be done than i thought now😅

BigRoy commented on client/ayon_openrv/api/networking.py:
```diff
@@ -0,0 +1,148 @@
+                break
+            
+            # get single message
+            (resp_type, resp_data) = self.receive_message()
```
These brackets are redundant (but maybe you just prefer them cosmetically?)
```suggestion
            resp_type, resp_data = self.receive_message()
```

BigRoy commented on client/ayon_openrv/startup/pkgs_source/ayon_menus/ayon_menus.py:
```diff
@@ -41,36 +49,38 @@ def install_openpype_to_host():
-                    ("_", None),  # separator
-                    ("Work Files...", self.workfiles, None, None),
-                ])
+            "",
+            None,
+            [
+                ("ayon_open_loader", self.load, "Opens AYON Loader."),
+                ("ayon_load_container", on_ayon_load_container, "Loads an AYON representation into the session.")
             ],
```
Even if positional arguments you should still be able to pass them by name, right?

BigRoy commented on client/ayon_openrv/plugins/load/load_mov.py:
```diff
@@ -52,7 +52,7 @@ def load(self, context, name=None, namespace=None, data=None):
         node = container["node"]
         repr_id = context["representation"]["id"]
-        context = get_representation_context(context["representation"])
+        context = get_representation_context(context["project"]["name"], context["representation"])
```
You don't need to do this ;) ... The `context` argument that `update` gets... is already the same as the `context` returned by `get_representation_context`.

So in essence:
```suggestion
```

tweak-wtf replied:
lol true 🤦

tweak-wtf commented:
i'd say for the first iteration this should be good.
@BigRoy I addressed all your suggestions (massive thanks again!) except for the one about openpype support. but i think i can just drop it since (as you said) openrv was never intended to be used with openpype

BigRoy commented on client/ayon_openrv/api/pipeline.py:
```diff
@@ -243,3 +257,18 @@ def get_containers():
             yield container
+
+
+def load_data(dataset=None, loader_type : str = None):
+    project_name = get_current_project_name()
```
This might benefit from a docstring. Especially because `data` is very generic and I have no idea what the `dataset` should be. It seems like `dataset` is supposed to be `representation_ids`. Also the function can't seem to work without any loader type specified or without representation_ids passed, as such the default `= None` values make no sense.

As such I'd expect this instead to be:

```suggestion
def load_representations(representation_ids: Iterable[str], loader_type: str):
    """Load multiple representation ids with a particular loader.
    
    Arguments:
        representation_ids: The representation ids to load.
        loader_type: The loader to use.
    """
    project_name = get_current_project_name()
```
etc.

tweak-wtf replied:
i pulled it out of `ayon_menus.py` since i'm calling it from `LoadContainerHandler` class. it's also being used in `data_loader()` in `ayon_menus.py`.
i didn't want to import from `startup/pkgs_source` in `LoadContainerHandler` and also tried to pull out as much logic from the menu generation.
that's why `load_data()` ended up in `pipeline.py`.

btw... `LoadContainerHandler` is my approach to refactor out the separation of LoaderType from `ayon_menus.py`.
Each custom event in RV could get one handler specific for that event type

BigRoy replied:
Good to know. Still think the function name and signature can be improved - but it being in `lib` then may make sense for the use cases here.

BigRoy commented on client/ayon_openrv/api/pipeline.py:
```diff
@@ -4,17 +4,31 @@
+try:
+    import rv
+except ImportError:
+    # NOTE: import fails because we're running outside RV
+    #       if we launch RV from TrayLoader 
+    #? how can i fix this
+    pass
```
I think you're best of putting the code you need to import elsewhere - e.g. the `networking` logic somewhere that does NOT import other dependencies of `ayon_rv` like `rv` itself if you want to use it outside of `rv`. 

In this case it seems to just the `networking.py` file.

I think for now you can get by with putting the `networking.py` outside of the `api.py` folder, e.g. next to `adddon.py` and you'd just import `ayon_rv.networking` so that you never import anything from `api`

jakubjezek001 commented:
I am not sure if this could be applicable for the networking but posting it here for inspiration anyway. 

We are using Websocket comunication mainly in TVPaint and AfterEffects. 

https://github.com/ynput/ayon-core/blob/a4fb13d476e244c188419d9ad9a5c185cd31857a/client/ayon_core/hosts/tvpaint/api/communication_server.py#L65

I believe the library @kalisp was using in AfterEffects might be more handy but I might be wrong.
https://github.com/ynput/ayon-core/blob/develop/client/ayon_core/hosts/aftereffects/api/ws_stub.py

We were already discussing it internally (@mkolar @m-u-r-p-h-y ) and the transformation of networking module could be done by us.

tweak-wtf commented:
Yeah also chatted with ondrej about the rpc.
I have a private branch in which i explored this approach a bit.
Would be very cool if you could give me some hints about implementing this.

tweak-wtf commented:
after chatting with @antirotor and @kalisp I'll merge in the latest `develop`,
expose amount of sleep after sending the `close` event as an environment variable AYON_RV_NETWORK_TIMEOUT`
and update the PR description to also show any pitfalls/caveats that are present

kalisp reviewed (approved):
I reworked `get_*_plugin` methods as we were loosing them inside of OpenRV.

Not totally sold on name `universal_loader` (which was `actions` before, which Kuba didn't like). Regular OpenRV load plugins must be in separate folder from `Play in OpenRV` as they are importing rv. Open for better name, maybe return back to `actions`??

BigRoy commented:
> I reworked `get_*_plugin` methods as we were loosing them inside of OpenRV.
> 
> Not totally sold on name `universal_loader` (which was `actions` before, which Kuba didn't like). Regular OpenRV load plugins must be in separate folder from `Play in OpenRV` as they are importing rv. Open for better name, maybe return back to `actions`??

I like the fact that at least it refers to loaders - since that's what it is. But I agree that `universal` is new terminology I haven't seen in AYON yet. I have heard of `global` before however. I wonder if `load_global` or even a subfolder of `load/global` would be better.

But by starting the folder with or inside of `load` then at least it appears closely related to that.

iLLiCiTiT commented on client/ayon_openrv/addon.py:
```diff
@@ -1,17 +1,42 @@
 import os
 
 from ayon_core.addon import AYONAddon, IHostAddon
+from ayon_core.addon.interfaces import IPluginPaths
```
```suggestion
from ayon_core.addon import AYONAddon, IHostAddon, IPluginPaths
```

iLLiCiTiT commented on client/ayon_openrv/addon.py:
```diff
@@ -1,17 +1,42 @@
+            # for Tray Loader
+            return [os.path.join(plugins_dir, "universal_load")]
+        # inside OpenRV only - import rv
+        return [os.path.join(plugins_dir, "load")]
```
```suggestion
        return [
            os.path.join(plugins_dir, "load"),
            os.path.join(plugins_dir, "universal_load")
        ]
```

iLLiCiTiT commented on client/ayon_openrv/networking.py:
```diff
@@ -0,0 +1,250 @@
+
+from ayon_openrv.version import __version__
+from ayon_openrv.addon import OpenRVAddon
+
+
```
```suggestion

```

iLLiCiTiT commented on client/ayon_openrv/networking.py:
```diff
@@ -0,0 +1,250 @@
+
+from ayon_core.lib.transcoding import (
+    IMAGE_EXTENSIONS, VIDEO_EXTENSIONS
+)
+
+from ayon_core.lib import Logger
+log = Logger.get_logger(__name__)
```
```suggestion
)
from ayon_core.lib import Logger

log = Logger.get_logger(__name__)
```

iLLiCiTiT commented on client/ayon_openrv/networking.py:
```diff
@@ -0,0 +1,250 @@
+
+            if resp_type == "PING":
+                self.sock.sendall("PONG 1 p".encode("utf-8"))
+            if resp_type == "PONG":
+                pass
+            if resp_type == "GREETING":
+                pass
+            if resp_type == "NEWGREETING":
+                pass
+            if resp_type == "NEWGREETING":
+                pass
```
```suggestion
```

iLLiCiTiT commented on client/ayon_openrv/networking.py:
```diff
@@ -0,0 +1,250 @@
+        log.debug(f"process message: {data = }")
+
+    def __process_events(self, process_return_only=False):
+        while True:
```
Shouldn't there be some `time.sleep(0.01)` in the loop to avoid 99% CPU usage?

iLLiCiTiT commented on client/ayon_openrv/version.py:
```diff
@@ -1,3 +1,3 @@
 # -*- coding: utf-8 -*-
 """Package declaring openrv addon version."""
-__version__ = "1.0.1-dev.1"
+__version__ = "1.1.0-dev.1"
```
```suggestion
__version__ = "1.0.1-dev.1"
```

iLLiCiTiT commented on package.py:
```diff
@@ -1,6 +1,6 @@
 name = "openrv"
 title = "OpenRV"
-version = "1.0.1-dev.1"
+version = "1.1.0-dev.1"
```
```suggestion
version = "1.0.1-dev.1"
```

iLLiCiTiT commented on server/version.py:
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
```
This files should not be there.

iLLiCiTiT reviewed (dismissed):
Few comments.

iLLiCiTiT commented:
> I like the fact that at least it refers to loaders - since that's what it is. But I agree that universal is new terminology I haven't seen in AYON yet. I have heard of global before however. I wonder if load_global or even a subfolder of load/global would be better.

That's in fact true, rather add subfolders `global` and `openrv` inside `load` folder.

tweak-wtf commented:
> > I like the fact that at least it refers to loaders - since that's what it is. But I agree that universal is new terminology I haven't seen in AYON yet. I have heard of global before however. I wonder if load_global or even a subfolder of load/global would be better.
> 
> That's in fact true, rather add subfolders `global` and `openrv` inside `load` folder.

i refactored that in https://github.com/ynput/ayon-openrv/pull/1/commits/8dea72f197f6dd5eb406f0e8c14c20634201520d
should be fine right?

kalisp commented:
Retested, still works.

iLLiCiTiT commented on client/ayon_openrv/addon.py:
```diff
@@ -1,17 +1,41 @@
+
+    def get_load_plugin_paths(self, host_name):
+        loaders_dir = os.path.join(OPENRV_ROOT_DIR, "plugins", "load")
+        if host_name != self.host_name:
+            # for Tray Loader
+            return [os.path.join(loaders_dir, "global")]
+        # inside OpenRV only - import rv
+        return [os.path.join(loaders_dir, "openrv")]
```
```suggestion
        # Global action
        paths = [os.path.join(loaders_dir, "global")]
        if host_name == self.host_name:
            # inside OpenRV only - import rv
            paths.append(os.path.join(loaders_dir, "openrv"))
        return paths
```

iLLiCiTiT replied:
Is it ok that the action will be available in Nuke, Maya, etc.?

kalisp replied:
Yeah, I think it is expected that you could play in OpenRV directly from Maya. (I tested it, it is working.)

iLLiCiTiT commented on client/ayon_openrv/plugins/load/global/play_in_rv.py:
```diff
@@ -0,0 +1,54 @@
+
+    product_types = {"*"}
+    representations = {"*"}
+    extensions = {
```
Can this use `IMAGE_EXTENSIONS` and `VIDEO_EXTENSIONS` from `ayon_core.lib.transcoding`?
```python
extensions = {
    ext.lstrip(".")
    for ext in IMAGE_EXTENSIONS| VIDEO_EXTENSIONS
}
```

iLLiCiTiT commented on client/ayon_openrv/networking.py:
```diff
@@ -0,0 +1,241 @@
+            )
+        self.event = event
+
+    def handle_event(self):
```
Shouldn't the event expect project name? What if I load from different project than current project?

BigRoy replied:
In that case - replace it with `load_with_repre_context` maybe? See [here](https://github.com/ynput/ayon-core/blob/e498b0aafe7ed27aca104fff520acab01ba7a880/client/ayon_core/pipeline/load/utils.py#L397-L409).

tweak-wtf commented:
@kalisp just made me aware that all my comments didn't show up on your sides. Sorry didn't know i had to submit my comments.
thought that `PENDING` flag meant the conversation was not resolved and thus pending 🤦

kalisp commented:
Merged, we can continue with next PRs.

Thanks for contribution! :)
